flax/training/checkpoints.py

```diff
@@ -313,3 +315,62 @@ def __call__(self, x):
     params_renamed[name] = convert_pre_linen(value)
 
   return core.freeze(params_renamed)
+
+
+class TrainState(struct.PyTreeNode):
+  """Simple train state for the common case with a single Optax optimizer.
+
+  Synopsis:
+
+    state = TrainState.create(
+        apply_fn=model.apply,
+        params=variables['params'],
+        tx=tx)
+    grad_fn = jax.grad(make_loss_fn(state.apply_fn))
+    for batch in data:
+      grads = grad_fn(state.params, batch)
+      state = state.update(grads=grads)
+
+  Note that you can easily extend this dataclass by subclassing it for storing
+  additional data (e.g. additional variable collections).
+
+  For more exotic usecases (e.g. multiple optimizers) it's probably best to
+  fork the class and modify it.
+
+  Attributes:
+    step: Counter starts at 0 and is incremented by every call to `.update()`.
+    apply_fn: Usually set to `model.apply()`. Kept in this dataclass for
+      convenience to have a shorter params list for `update_step()`.
+    tx: An Optax gradient transformation.
+    opt_state: The state for `tx`.
+  """
+  step: int
+  apply_fn: Callable = struct.field(pytree_node=False)
+  params: core.FrozenDict[str, Any]
+  tx: optax.GradientTransformation = struct.field(pytree_node=False)
+  opt_state: optax.OptState
+
+  def update(self, *, grads, **kwargs):
+    """Updates `step`, `params`, `opt_state` and `**kwargs` in return value."""
```

Comment:
Can we clarify in the docstring that actually takes a gradient step, and in addition updates other kwards? The current docstring doesn't really say much. And/or maybe even the name? Like apply_gradients?

Reply:
Renamed & expanded documentation to make this clearer.

Note that we used to write `.apply_gradient()` (note the missing `s`) with `flax.optim.Opimizer` - but I think `.apply_gradients()` is better because we usually say `params` and `grads` when talking about entire pytrees of parameters/gradients. Also, we we're not trying to be compatible with `flax.optim.Optimizer` so I think it's fine to add the `s` here...

Reply:
Agreed.